### ISSUE TITLE
Stuttering and stalling with Gemini and G11

### ISSUE THREAD

**Issue by bgottula** [seeing-things/track#94]
Sometimes for no apparent reason when attempting to track with Gemini the G11 mount will stutter violently, sometimes badly enough to induce motor stalls. This makes practical tracking impossible and is probably hard on the motors. There is no known reliable way to recover from this issue, including power cycling Gemini. Yet, somehow, the issue will then spontaneously vanish and smooth performance returns. This has happened on multiple occasions while testing the software indoors and as of 8 June 2018 has also cost us a 80-degree elevation ISS pass when it started happening just when the hybrid_track program was started up to track ISS.

Some ideas for what might be going on:
- Our software (either point or track) is doing something wrong. Malformed commands, commanding erratic slew rates with high accelerations, long pauses between slew commands, etc.
- Gemini firmware has a bug or has some behavior we don't understand. On 8 June while attempting to track ISS I tried to recover from this problem by power-cycling Gemini and that did not seem to help which should rule out some classes of bugs.
- Gemini software/firmware is fighting our software for control of the slew rate. It is not entirely clear to me how we know what is in control of the slew rate at any given moment. For example, if we leave the tracking mode as "sidereal," does that mean Gemini is going to periodically try to override our divisor commands?

Things that have been ruled out:

- Errant detections in the guidescope camera. This behavior has been observed when using blind_track, when the camera is not involved at all.
- Hardware problem, such as damage to the servo motors, gears, or wiring. This almost certainly is not the case because when the mount is symptomatic, the problem vanishes when performing a goto using the hand controller. The problem also goes away when slewing in the problem direction after going back the other way and then returning.
- Hardware dynamics (gears too loose or too tight, not lubricated properly, unbalanced load). This seems unlikely based on the symptoms unless something really counter-intuitive is going on, like lubrication being redistributed after moving one direction in RA and thereby reducing friction on subsequent passes of that part of the gear. I took pains to test on 9 June with very minimal load on the RA axis. Inspection of the worm did not reveal anything concerning. The drivetrain upstream of the worm is fully exonerated because it turns multiple full revolutions with the issue and without the issue.
- Errant detection of RA axis limit. I enabled debug messages in the gamepad program and did not see any evidence that the axis limit was tripped when it should not have been.

As part of the resolution of this issue, it might be prudent to detect and handle motor stalls (if possible). When Gemini detects a motor stall the following things seem to happen:

- Motion on that axis stops (not sure about other axis)
- Beep sound
- Get NACKs on UDP interface for some stupid reason
- Track program is unaware and keeps trying to command potentially high slew rates which causes further motor stalls

The appropriate response is probably to reset the loop filter integrator on the affected axis to 0 to sync up with the fact that motion on the axis has stopped. This will hopefully allow that axis to accelerate smoothly back up to full rate without causing further stalls.

**Comment by bgottula** [comment]
Some observations based on testing on 9 June:

- Both today and on the 8th, I have only observed the problem affecting the RA axis. The DEC axis seems unaffected.
- The problem only seems to affect the RA axis when moving in one direction but not the other. The problem direction is when the RA axis is moving clockwise as seen when standing on the north side of the mount. This is the same direction the mount moves to track sidereal motion.
- The stuttering seems to occur at all slew rates.
- The stutter seems to disappear after traversing a certain span of physical position the first time. In other words, after slewing with stutter in the RA axis to a certain point, then backing up several degrees in RA and then proceeding again in the problem direction, the motion will be smooth in the problem direction until reaching the same point in hour angle. Then stuttering resumes. After slewing with stuttering all the way to the RA limit, then it is possible to slew back and forth in RA across the entire range completely smoothly.

**Comment by bgottula** [comment]
I have started performing additional experimentation. This time rather than using higher level control software or the gamepad_control program I am invoking the slew() method on the LosmandyGeminiMount class in the track package. This should (I think) only be sending a single native Gemini command (451 or 452) to the mount per invocation. I have observed the following:

* Additional confirmation that only the RA axis seems to be affected.
* After setting the slew rate on the RA axis to a value such as +1 or -1, the mount will slew for some time and then stop. Inspection of the Gemini State webpage shows that the slew rate has not actually stopped but has instead changed to a very slow rate. The slow slew rate always continues in the same direction as the original slew. If the slew rate is explicitly set to 0, the mount motion will actually come to a stop.
* The issue is present even when the tracking mode is set to "Comet/User Defined". René had recommended using Comet tracking mode to make sure our divisors are preserved.

These observations suggest that the stuttering behavior is in fact resulting from something in the Gemini firmware setting the RA divisor to values that are in conflict with the values we are trying to set using our software. I attempted to confirm this by reading the divisors but the native 451 and 452 commands are evidently write-only.

I also ruled out the hand controller by performing the same experiments with the hand controller disconnected from Gemini 2. I did not observe any change in behavior with the hand controller disconnected.

**Comment by bgottula** [comment]
I think I figured it out. I saved my SRAM settings to a file and then performed a factory reset of the configuration. After rebooting, the problem behavior went away. I did a diff of the settings in my configuration file against the factory reset file. I found a suspicious setting that was enabled only in my file named "Activate PEC at boot". I have mostly ignored the PEC settings and never intended to enable PEC of any kind, so I was surprised that it was enabled.

After the factory reset, I enabled only the the "Activate PEC at boot" setting and left everything else the same to make sure this was the real setting of importance and not a red herring. After a reboot, the problem behavior returned. I think this confirms with fairly high confidence that this is the source of the problem.

The outstanding mystery is how this feature became enabled. My guess is that I either enabled it by mistake using the hand controller or the web interface or that it was enabled by a malformed serial command. This is perhaps good motivation for adding checks on Gemini configuration during startup in track programs.

**Comment by bgottula** [comment]
During observing session on 7 August 2018 this issue reoccurred, even though the "Activate PEC at boot" setting was disabled in the stored configuration (unless I made a mistake).

Another theory is that this issue also manifests when a model is loaded, either from a file or by performing an alignment.

**Comment by bgottula** [comment]
During unrelated testing on 3 Aug 2019 the stuttering was happening frequently, with all the same symptoms as usual (RA axis only, usually only in one direction, and only over range of angles not traversed yet). I found to my surprise that PEC was active and "Activate PEC at Boot" was checked in the hand controller menus. I did A/B tests with PEC enabled and disabled via the hand controller and confirmed again that disabling PEC made the problem go away.

I wonder if Gemini is somehow re-enabling PEC without being asked. I'm pretty confident that I have never intentionally enabled it. Other settings like IP addresses were preserved.
